CDAP-14098 Fix resource leak in preview

anwar6953 opened the pull request:
https://issues.cask.co/browse/CDAP-14098
Make LevelDBTableService into an AbstractIdleService, so that it has Service lifecycle of startup and shutdown. Close All of its LevelDB DB object upon shutdown.

Commit:
CDAP-14098 Make LevelDBTableService into an AbstractIdleService, so that it has Service lifecycle of startup and shutdown. Close All of its LevelDB DB object upon shutdown.
Fix LevelDBTableTest, which was not performing some assertions that the comments implied were being asserted.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -59,7 +63,7 @@
   private WriteOptions writeOptions;
 
-  private final ConcurrentMap<String, DB> tables = Maps.newConcurrentMap();
+  private ConcurrentMap<String, DB> tables;
```
Why need to change?

chtyim replied:
Also, shouldn't all methods checks if this service is shutdown or not?

anwar6953 replied:
"Why need to change?"
Same reason as https://github.com/caskdata/cdap/pull/10535#discussion_r210696537

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -98,6 +102,18 @@ public void clearTables() {
+
+  @Override
+  protected void shutDown() {
+    for (DB entries : tables.values()) {
```
Should also clear the cached `tables`.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -50,7 +54,7 @@
  */
 @Singleton
-public class LevelDBTableService {
+public class LevelDBTableService extends AbstractIdleService {
```
Why a service? Shouldn't it be a `Closeable`?

anwar6953 replied:
I made it a service to first find out what are all the places we need to control its Lifecycle.
I found that we need to do it in StandaloneMain and in PreviewRunner.

I'll change it to a Closeable now.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -59,7 +60,7 @@
   private WriteOptions writeOptions;
 
-  private final ConcurrentMap<String, DB> tables = Maps.newConcurrentMap();
+  private ConcurrentMap<String, DB> tables = Maps.newConcurrentMap();
```
Should remain as `final`.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -95,9 +96,16 @@ public void setConfiguration(CConfiguration config) throws IOException {
   }
 
+  public void close() {
+    clearTables();
```
Shouldn't all methods should check if this table service is closed? and if it is closed, throws exception?

anwar6953 replied:
Done.

anew commented:
Should other classes that use LevelDBTableService also close it? 
Mostly there are test classes that use it: LevelDBTableTest, LocalQueueTest, LevelDBQueueTest, maybe others.

anwar6953 commented:
@anew addressed comments.
Also fixed `LevelDBTableTest`, where the comments implied it was performing some some assertion, but the assertion was not actually being made.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableTest.java:
```diff
@@ -106,10 +106,11 @@ public void testTablesSurviveAcrossRestart() throws Exception {
     service.clearTables();
     for (String tableName : tableNames) {
-      service.list().contains(tableName);
+      Preconditions.checkState(service.list().contains(
```
good one :)

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/kv/LevelDBKVTableTest.java:
```diff
@@ -38,15 +39,20 @@
+
+  @After
+  public void clear() {
+    levelDBTableService.clearTables();
```
don't you rather want to call close()?

anwar6953 replied:
Since the service is like a singleton object, I don't want to call close, which makes the service unusable after closing it.
`clearTables` is very similar.

anew replied:
got it

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableTest.java:
```diff
@@ -106,10 +106,11 @@ public void testTablesSurviveAcrossRestart() throws Exception {
+        PrefixedNamespaces.namespace(cConf, CONTEXT1.getNamespaceId(), tableName)));
     }
   }
 }
```
My previous comment I meant that this test should have an @After clause that calls levelDBTableService.close(). And same for the other tests.

anwar6953 replied:
Since the service is like a singleton object, I don't want to call close, which makes the service unusable after closing it.
`clearTables` is very similar.

anew reviewed (commented):
two more comments

anwar6953 commented:
Build is green: https://builds.cask.co/browse/CDAP-DUT6581-5

anwar6953 commented:
Squashed commits; merging.
